apps/s_client.c

```diff
@@ -2257,6 +2259,87 @@ int s_client_main(int argc, char **argv)
             }
         }
         break;
+    case PROTO_MYSQL:
+        {
+            /* SSL request packet */
+            static const unsigned char ssl_req[] = {
+                /* payload_length,   sequence_id */
+                   0x20, 0x00, 0x00, 0x01,
+                /* payload */
+                /* capability flags, CLIENT_SSL always set */
+                   0x85, 0xae, 0x7f, 0x00,
+                /* max-packet size */
+                   0x00, 0x00, 0x00, 0x01,
+                /* character set */
+                   0x21,
+                /* string[23] reserved (all [0]) */
+                   0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+                   0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+                   0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00
+            };
+            int bytes = 0;
+            int ssl_flg = 0x800;
+            int pos;
+
+            /* Receiving Initial Handshake packet. */
+            bytes = BIO_read(sbio, sbuf, BUFSIZZ);
+            /* Packet length[3], Packet number[1] + minimum payload[17] */
+            if (bytes < 21) {
+                BIO_printf(bio_err, "MySQL packet too short.\n");
+                goto shut;
+            /* header[4] + payload(Assume max 255 bytes) */
+            } else if (bytes > (4 + 0xFF)) {
+                BIO_printf(bio_err, "MySQL packet too long.\n");
+                goto shut;
+            } else if ((unsigned char)sbuf[1] != 0 ||
+                       (unsigned char)sbuf[2] != 0) {
+                BIO_printf(bio_err,
+                           "Payload over 256 bytes is not supported.\n");
+                goto shut;
+            } else if (bytes != ((unsigned char)sbuf[0] + 4)) {
+                BIO_printf(bio_err, "MySQL packet length does not match.\n");
+                goto shut;
+            /* protocol version[1] */
+            } else if (sbuf[4] != 0xA) {
+                BIO_printf(bio_err,
+                           "Only MySQL protocol version 10 is supported.\n");
+                goto shut;
+            }
+
+            pos = 5;
+            /* server version[string+NULL] */
+            for (;;) {
+                if (sbuf[pos] == '\0') {
+                    break;
+                } else if (pos > bytes) {
+                    BIO_printf(bio_err,
+                              "Cannot confirm server version. "
+                              "MySQL server handshake packet is broken.\n");
+                    goto shut;
+                }
+                pos++;
+            }
+
+            /* conn id[4] + SALT[8] + NULL */
```

Comment:
While it does work, it makes it less readable [and harder to review]. I mean when you see if (pos >= bytes) you recognize the pattern as it is, scanning for *a* null-terminated string that doesn't overstep *a* buffer. Bottom line is that I'd really prefer to see separate check for enough remaining data as originally suggested.

Another readability thing is that upon exit from loop pos points at '\0'. This is kind of counter-intuitive, as one would expect it to point *past* "processed" data. This is easily fixed by using `packet[pos++] == '\0'`. And it would also be appropriate to explicitly advance pos so that references to capability vector would be `packet[pos] + (packet[pos + 1] << 8)` or to summarize
```
    } else if (packet[pos++] == '\0') {
        break;
    }
}
if (pos + 15 > bytes) {
    bad thing;
}
pos += 12; /* skip over conn id[4] + SALT[8] */
if (packet[pos++] != '\0') { /* verify filler */
    bad thing;
}
if (!((packet[pos] + (packet[pos + 1] << 8)) & ssl_flg)) {
    bad thing;
}
```